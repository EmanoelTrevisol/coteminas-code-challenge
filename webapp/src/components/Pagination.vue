<template>
  <div class="pagination">
    <div class="per-page-options">
      <select
        name="per-page"
        id="per-page"
        @change="(evt) => $emit('update:perPage', parseInt(evt.target.value))"
      >
        <option
          v-for="option in perPageOptions"
          :value="option"
          :key="option"
          :selected="option === perPage"
        >
          {{ option }} produtos por p√°gina
        </option>
      </select>
    </div>
    <div class="page-options">
      <span
        class="first-page"
        :class="{ disabled: !canGoToPreviousPage() }"
        @click="firstPage"
      >
        <font-awesome-icon icon="angle-double-left"></font-awesome-icon>
      </span>
      <span
        class="previous-page"
        :class="{ disabled: !canGoToPreviousPage() }"
        @click="previousPage"
      >
        <font-awesome-icon icon="angle-left"></font-awesome-icon>
      </span>
      <span
        v-for="pageOption in pageOptions.previous"
        :key="pageOption"
        class="option"
        @click="onPageClick(pageOption)"
      >
        {{ pageOption }}
      </span>
      <span class="option selected">
        {{ currentPage }}
      </span>
      <span
        v-for="pageOption in pageOptions.next"
        :key="pageOption"
        class="option"
        @click="onPageClick(pageOption)"
      >
        {{ pageOption }}
      </span>
      <span
        class="next-page"
        :class="{ disabled: !canGoToNextPage() }"
        @click="nextPage"
      >
        <font-awesome-icon icon="angle-right"></font-awesome-icon>
      </span>
      <span
        class="last-page"
        :class="{ disabled: !canGoToNextPage() }"
        @click="lastPage"
      >
        <font-awesome-icon icon="angle-double-right"></font-awesome-icon>
      </span>
    </div>
  </div>
</template>

<script>
export default {
  props: {
    total: {
      type: Number,
      required: true,
    },
    currentPage: {
      type: Number,
      default: 1,
    },
    perPage: {
      type: Number,
      default: 10,
    },
  },
  data() {
    return {
      perPageOptions: [10, 25, 50, 100],
    };
  },
  computed: {
    totalPages() {
      return (
        (this.total && this.perPage && Math.ceil(this.total / this.perPage)) ||
        0
      );
    },
    pageOptions() {
      const previous = [];
      const next = [];

      const { prevLength, nextLength } = this.getOptionsLength();

      for (
        let i = this.currentPage - 1;
        i >= this.currentPage - prevLength;
        i--
      ) {
        if (i < 1) continue;
        previous.unshift(i);
      }

      for (
        let i = this.currentPage + 1;
        i <= this.currentPage + nextLength;
        i++
      ) {
        if (i > this.totalPages) continue;
        next.push(i);
      }

      return {
        previous,
        next,
      };
    },
  },
  methods: {
    getOptionsLength() {
      const opts = {};
      const leftPagesToLast = this.totalPages - this.currentPage;

      if (this.currentPage < 4) {
        opts.prevLength = this.currentPage - 1;
        opts.nextLength = 6 - this.currentPage;
      } else if (leftPagesToLast < 4) {
        opts.prevLength = 6 - leftPagesToLast;
        opts.nextLength = leftPagesToLast;
      } else {
        opts.prevLength = 3;
        opts.nextLength = 3;
      }

      return opts;
    },
    onPageClick(pageNumber) {
      this.$emit('update:currentPage', pageNumber);
    },
    canGoToNextPage() {
      return !!this.pageOptions.next.length;
    },
    canGoToPreviousPage() {
      return !!this.pageOptions.previous.length;
    },
    nextPage(evt) {
      if (this.canGoToNextPage()) {
        this.$emit('update:currentPage', this.currentPage + 1);
      } else {
        evt.stopPropagation();
        evt.preventDefault();
      }
    },
    lastPage(evt) {
      if (this.canGoToNextPage()) {
        this.$emit('update:currentPage', this.totalPages);
      } else {
        evt.stopPropagation();
        evt.preventDefault();
      }
    },
    previousPage(evt) {
      if (this.canGoToPreviousPage()) {
        this.$emit('update:currentPage', this.currentPage - 1);
      } else {
        evt.stopPropagation();
        evt.preventDefault();
      }
    },
    firstPage(evt) {
      if (this.canGoToPreviousPage()) {
        this.$emit('update:currentPage', 1);
      } else {
        evt.stopPropagation();
        evt.preventDefault();
      }
    },
  },
};
</script>

<style lang="stylus" scoped>
.pagination {
  display: flex;
  justify-content: space-between;
  align-items: center;
  width: 100%;
  height: 40px;

  .per-page-options {
    select {
      height: 40px;
      padding: 5px 10px;
      font-size: 1.05em;
      font-family: $ff-roboto;
      font-weight: 300;
      border-radius: 5px;
      border-color: $medium;
    }
  }

  .page-options {
    display: flex;
    justify-content: space-evenly;
    align-items: center;

    .option, .first-page, .previous-page, .next-page, .last-page {
      height: 30px
      padding: 5px 10px;
      margin: 2px;
      font-weight: 300;
      font-size: 1.1em;
      cursor: pointer;
      border-radius: 5px;

      &.disabled {
        color: lighten($medium, 30%);
        cursor: default;
      }

      &:hover {
        background-color: $light;
      }
    }

    .option.selected {
      border: 1px solid $light;
    }
  }
}
</style>
