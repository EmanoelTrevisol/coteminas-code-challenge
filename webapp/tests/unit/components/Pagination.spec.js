import { shallowMount } from '@vue/test-utils';
import Pagination from '@/components/Pagination';

describe('Pagination.vue', () => {
  describe('render component', () => {
    it('renders Pagination component', () => {
      const wrapper = shallowMount(Pagination);

      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('computes correct options based on props', () => {
    const build = (data) => {
      const wrapper = shallowMount(Pagination, { propsData: data || {} });

      const options = wrapper.findAll('.option').wrappers;

      const numOptions = options.map((wrp) => wrp.text());

      return { wrapper, numOptions };
    };

    it('displays correct current page', () => {
      const currentPage = 5;
      const { wrapper } = build({
        currentPage,
        total: 103,
        perPage: 10,
      });

      const selected = wrapper.find('.option.selected');

      expect(selected.text()).toBe(currentPage.toString());
    });

    describe('page options depending on currentPage', () => {
      it('current: 1, total: 103, perPage: 10', () => {
        const currentPage = 1;
        const { numOptions } = build({
          currentPage,
          total: 103,
          perPage: 10,
        });

        expect(numOptions).toStrictEqual(['1', '2', '3', '4', '5', '6']);
      });

      it('current: 2, total: 103, perPage: 10', () => {
        const currentPage = 2;
        const { numOptions } = build({
          currentPage,
          total: 103,
          perPage: 10,
        });

        expect(numOptions).toStrictEqual(['1', '2', '3', '4', '5', '6']);
      });

      it('current: 3, total: 103, perPage: 10', () => {
        const currentPage = 3;
        const { numOptions } = build({
          currentPage,
          total: 103,
          perPage: 10,
        });

        expect(numOptions).toStrictEqual(['1', '2', '3', '4', '5', '6']);
      });

      it('current: 4, total: 103, perPage: 10', () => {
        const currentPage = 4;
        const { numOptions } = build({
          currentPage,
          total: 103,
          perPage: 10,
        });

        expect(numOptions).toStrictEqual(['1', '2', '3', '4', '5', '6', '7']);
      });

      it('current: 5, total: 103, perPage: 10', () => {
        const currentPage = 5;
        const { numOptions } = build({
          currentPage,
          total: 103,
          perPage: 10,
        });

        expect(numOptions).toStrictEqual(['2', '3', '4', '5', '6', '7', '8']);
      });

      it('current: 6, total: 103, perPage: 10', () => {
        const currentPage = 6;
        const { numOptions } = build({
          currentPage,
          total: 103,
          perPage: 10,
        });

        expect(numOptions).toStrictEqual(['3', '4', '5', '6', '7', '8', '9']);
      });

      it('current: 7, total: 103, perPage: 10', () => {
        const currentPage = 7;
        const { numOptions } = build({
          currentPage,
          total: 103,
          perPage: 10,
        });

        expect(numOptions).toStrictEqual(['4', '5', '6', '7', '8', '9', '10']);
      });

      it('current: 8, total: 103, perPage: 10', () => {
        const currentPage = 8;
        const { numOptions } = build({
          currentPage,
          total: 103,
          perPage: 10,
        });

        expect(numOptions).toStrictEqual(['5', '6', '7', '8', '9', '10', '11']);
      });

      it('current: 9, total: 103, perPage: 10', () => {
        const currentPage = 9;
        const { numOptions } = build({
          currentPage,
          total: 103,
          perPage: 10,
        });

        expect(numOptions).toStrictEqual(['5', '6', '7', '8', '9', '10', '11']);
      });

      it('current: 10, total: 103, perPage: 10', () => {
        const currentPage = 10;
        const { numOptions } = build({
          currentPage,
          total: 103,
          perPage: 10,
        });

        expect(numOptions).toStrictEqual(['5', '6', '7', '8', '9', '10', '11']);
      });

      it('current: 11, total: 103, perPage: 10', () => {
        const currentPage = 11;
        const { numOptions } = build({
          currentPage,
          total: 103,
          perPage: 10,
        });

        expect(numOptions).toStrictEqual(['5', '6', '7', '8', '9', '10', '11']);
      });
    });

    describe('page options depending on perPage', () => {
      it('current: 1, total: 101, perPage: 10', () => {
        const currentPage = 1;
        const { numOptions } = build({
          currentPage,
          total: 101,
          perPage: 10,
        });

        expect(numOptions).toStrictEqual(['1', '2', '3', '4', '5', '6']);
      });

      it('current: 1, total: 101, perPage: 25', () => {
        const currentPage = 1;
        const { numOptions } = build({
          currentPage,
          total: 101,
          perPage: 25,
        });

        expect(numOptions).toStrictEqual(['1', '2', '3', '4', '5']);
      });

      it('current: 1, total: 101, perPage: 50', () => {
        const currentPage = 1;
        const { numOptions } = build({
          currentPage,
          total: 101,
          perPage: 50,
        });

        expect(numOptions).toStrictEqual(['1', '2', '3']);
      });

      it('current: 1, total: 101, perPage: 100', () => {
        const currentPage = 1;
        const { numOptions } = build({
          currentPage,
          total: 101,
          perPage: 100,
        });

        expect(numOptions).toStrictEqual(['1', '2']);
      });
    });

    describe('page options depending on perPage and currentPage', () => {
      it('current: 1, total: 101, perPage: 25', () => {
        const currentPage = 1;
        const { numOptions } = build({
          currentPage,
          total: 101,
          perPage: 25,
        });

        expect(numOptions).toStrictEqual(['1', '2', '3', '4', '5']);
      });

      it('current: 2, total: 101, perPage: 25', () => {
        const currentPage = 2;
        const { numOptions } = build({
          currentPage,
          total: 101,
          perPage: 25,
        });

        expect(numOptions).toStrictEqual(['1', '2', '3', '4', '5']);
      });

      it('current: 3, total: 101, perPage: 25', () => {
        const currentPage = 3;
        const { numOptions } = build({
          currentPage,
          total: 101,
          perPage: 25,
        });

        expect(numOptions).toStrictEqual(['1', '2', '3', '4', '5']);
      });

      it('current: 4, total: 101, perPage: 25', () => {
        const currentPage = 4;
        const { numOptions } = build({
          currentPage,
          total: 101,
          perPage: 25,
        });

        expect(numOptions).toStrictEqual(['1', '2', '3', '4', '5']);
      });

      it('current: 5, total: 101, perPage: 25', () => {
        const currentPage = 5;
        const { numOptions } = build({
          currentPage,
          total: 101,
          perPage: 25,
        });

        expect(numOptions).toStrictEqual(['1', '2', '3', '4', '5']);
      });

      it('current: 1, total: 101, perPage: 50', () => {
        const currentPage = 1;
        const { numOptions } = build({
          currentPage,
          total: 101,
          perPage: 50,
        });

        expect(numOptions).toStrictEqual(['1', '2', '3']);
      });

      it('current: 2, total: 101, perPage: 50', () => {
        const currentPage = 2;
        const { numOptions } = build({
          currentPage,
          total: 101,
          perPage: 50,
        });

        expect(numOptions).toStrictEqual(['1', '2', '3']);
      });

      it('current: 3, total: 101, perPage: 50', () => {
        const currentPage = 3;
        const { numOptions } = build({
          currentPage,
          total: 101,
          perPage: 50,
        });

        expect(numOptions).toStrictEqual(['1', '2', '3']);
      });

      it('current: 1, total: 35, perPage: 10', () => {
        const currentPage = 1;
        const { numOptions } = build({
          currentPage,
          total: 35,
          perPage: 10,
        });

        expect(numOptions).toStrictEqual(['1', '2', '3', '4']);
      });

      it('current: 2, total: 35, perPage: 10', () => {
        const currentPage = 2;
        const { numOptions } = build({
          currentPage,
          total: 35,
          perPage: 10,
        });

        expect(numOptions).toStrictEqual(['1', '2', '3', '4']);
      });

      it('current: 3, total: 35, perPage: 10', () => {
        const currentPage = 3;
        const { numOptions } = build({
          currentPage,
          total: 35,
          perPage: 10,
        });

        expect(numOptions).toStrictEqual(['1', '2', '3', '4']);
      });

      it('current: 4, total: 35, perPage: 10', () => {
        const currentPage = 4;
        const { numOptions } = build({
          currentPage,
          total: 35,
          perPage: 10,
        });

        expect(numOptions).toStrictEqual(['1', '2', '3', '4']);
      });

      it('current: 1, total: 45, perPage: 10', () => {
        const currentPage = 1;
        const { numOptions } = build({
          currentPage,
          total: 45,
          perPage: 10,
        });

        expect(numOptions).toStrictEqual(['1', '2', '3', '4', '5']);
      });

      it('current: 3, total: 45, perPage: 10', () => {
        const currentPage = 2;
        const { numOptions } = build({
          currentPage,
          total: 45,
          perPage: 10,
        });

        expect(numOptions).toStrictEqual(['1', '2', '3', '4', '5']);
      });

      it('current: 3, total: 45, perPage: 10', () => {
        const currentPage = 3;
        const { numOptions } = build({
          currentPage,
          total: 45,
          perPage: 10,
        });

        expect(numOptions).toStrictEqual(['1', '2', '3', '4', '5']);
      });

      it('current: 4, total: 45, perPage: 10', () => {
        const currentPage = 4;
        const { numOptions } = build({
          currentPage,
          total: 45,
          perPage: 10,
        });

        expect(numOptions).toStrictEqual(['1', '2', '3', '4', '5']);
      });

      it('current: 5, total: 45, perPage: 10', () => {
        const currentPage = 5;
        const { numOptions } = build({
          currentPage,
          total: 45,
          perPage: 10,
        });

        expect(numOptions).toStrictEqual(['1', '2', '3', '4', '5']);
      });
    });

    describe('show perPage options correctly', () => {
      it('should display perPage select options', () => {
        const options = [
          { value: '10', text: '10 produtos por página' },
          { value: '25', text: '25 produtos por página' },
          { value: '50', text: '50 produtos por página' },
          { value: '100', text: '100 produtos por página' },
        ];

        const { wrapper } = build({
          currentPage: 1,
          total: 45,
          perPage: 10,
        });

        const select = wrapper.find('select');

        const [opt0, opt1, opt2, opt3] = select.element.options;

        expect(opt0.text).toBe(options[0].text);
        expect(opt0.value).toBe(options[0].value);

        expect(opt1.text).toBe(options[1].text);
        expect(opt1.value).toBe(options[1].value);

        expect(opt2.text).toBe(options[2].text);
        expect(opt2.value).toBe(options[2].value);

        expect(opt3.text).toBe(options[3].text);
        expect(opt3.value).toBe(options[3].value);
      });

      it('should have 25 perPage selected option', () => {
        const { wrapper } = build({
          currentPage: 1,
          total: 45,
          perPage: 25,
        });

        const select = wrapper.find('select');

        expect(select.find('option:checked').element.value).toBe('25');
        expect(select.find('option:checked').element.text).toBe(
          '25 produtos por página'
        );
      });
    });

    describe('emits events to the parent component', () => {
      it('should select 50 perPage option and emit event', async () => {
        const { wrapper } = build({
          currentPage: 1,
          total: 45,
          perPage: 10,
        });

        const select = wrapper.find('select');

        await select
          .findAll('option')
          .at(2)
          .setSelected();
        await wrapper.setProps({
          perPage: 50,
        });

        expect(select.find('option:checked').element.value).toBe('50');
        expect(select.find('option:checked').element.text).toBe(
          '50 produtos por página'
        );

        expect(wrapper.emitted('update:perPage').length).toBe(1);
        expect(wrapper.emitted('update:perPage')[0]).toStrictEqual([50]);
      });

      it('should emit update:currentPage on item click', async () => {
        const { wrapper } = build({
          currentPage: 1,
          total: 45,
          perPage: 10,
        });

        const newPage = wrapper.findAll('.option').at(1);
        expect(newPage.text()).toBe('2');

        await newPage.trigger('click');
        await wrapper.setProps({ currentPage: 2 });

        const currentPageEmitted = wrapper.emitted('update:currentPage');

        expect(wrapper.find('.option.selected').text()).toBe('2');
        expect(currentPageEmitted.length).toBe(1);
        expect(currentPageEmitted[0]).toStrictEqual([2]);
      });

      it('should emit update:currentPage on first page click', async () => {
        const { wrapper } = build({
          currentPage: 3,
          total: 45,
          perPage: 10,
        });

        const firstPage = wrapper.find('.first-page');

        await firstPage.trigger('click');
        await wrapper.setProps({
          currentPage: 1,
        });

        const currentPageEmitted = wrapper.emitted('update:currentPage');

        expect(wrapper.find('.option.selected').text()).toBe('1');
        expect(firstPage.classes('disabled')).toBe(true);
        expect(currentPageEmitted.length).toBe(1);
        expect(currentPageEmitted[0]).toStrictEqual([1]);
      });

      it('should emit update:currentPage on last page click', async () => {
        const { wrapper } = build({
          currentPage: 3,
          total: 45,
          perPage: 10,
        });

        const lastPage = wrapper.find('.last-page');

        await lastPage.trigger('click');
        await wrapper.setProps({
          currentPage: 5,
        });

        const currentPageEmitted = wrapper.emitted('update:currentPage');

        expect(wrapper.find('.option.selected').text()).toBe('5');
        expect(lastPage.classes('disabled')).toBe(true);
        expect(currentPageEmitted.length).toBe(1);
        expect(currentPageEmitted[0]).toStrictEqual([5]);
      });

      it('should emit update:currentPage on previous page click', async () => {
        const { wrapper } = build({
          currentPage: 3,
          total: 45,
          perPage: 10,
        });

        const previous = wrapper.find('.previous-page');

        await previous.trigger('click');
        await wrapper.setProps({
          currentPage: 2,
        });

        const currentPageEmitted = wrapper.emitted('update:currentPage');

        expect(wrapper.find('.option.selected').text()).toBe('2');
        expect(previous.classes('disabled')).toBe(false);
        expect(currentPageEmitted.length).toBe(1);
        expect(currentPageEmitted[0]).toStrictEqual([2]);
      });

      it('should emit update:currentPage on next page click', async () => {
        const { wrapper } = build({
          currentPage: 3,
          total: 45,
          perPage: 10,
        });

        const next = wrapper.find('.next-page');

        await next.trigger('click');
        await wrapper.setProps({
          currentPage: 4,
        });

        const currentPageEmitted = wrapper.emitted('update:currentPage');

        expect(wrapper.find('.option.selected').text()).toBe('4');
        expect(next.classes('disabled')).toBe(false);
        expect(currentPageEmitted.length).toBe(1);
        expect(currentPageEmitted[0]).toStrictEqual([4]);
      });
    });
  });
});
