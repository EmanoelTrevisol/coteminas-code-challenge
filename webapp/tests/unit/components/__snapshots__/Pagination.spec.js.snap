// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination.vue render component renders Pagination component 1`] = `
<div class="pagination">
  <div class="per-page-options"><select name="per-page" id="per-page">
      <option value="10">
        10 produtos por p치gina
      </option>
      <option value="25">
        25 produtos por p치gina
      </option>
      <option value="50">
        50 produtos por p치gina
      </option>
      <option value="100">
        100 produtos por p치gina
      </option>
    </select></div>
  <div class="page-options"><span class="first-page disabled"><font-awesome-icon icon="angle-double-left"></font-awesome-icon></span> <span class="previous-page disabled"><font-awesome-icon icon="angle-left"></font-awesome-icon></span> <span class="option selected">
      1
    </span> <span class="next-page disabled"><font-awesome-icon icon="angle-right"></font-awesome-icon></span> <span class="last-page disabled"><font-awesome-icon icon="angle-double-right"></font-awesome-icon></span></div>
</div>
`;
